File: c++/leetcode/69. Sqrt(x).cpp
class Solution
{
public:
  int mySqrt(int x)
  {
    int end = x;
    int start = 0;

    while (start <= end)
    {
      int mid = start + ((end - start) / 2);
      long sqr = pow(mid, 2);

      if (x == sqr)
        return mid;

      if (sqr < x)
        start = ++mid;
      else
        end = --mid;
    }

    return end;
  }
};

File: javascript/leetcode/69. Sqrt(x).js
/**
 * @param {number} x
 * @return {number}
 */
var mySqrt = function (x) {
  let end = x;
  let start = 0;

  while (start <= end) {
    let mid = start + Math.floor((end - start) / 2);
    const sqr = Math.pow(mid, 2);

    if (x === sqr) return mid;

    if (sqr < x) start = ++mid;
    else end = --mid;
  }

  return end;
};
